### ISSUE TITLE
Termux looks for a hosting service

### ISSUE THREAD

**Issue by xeffyr** [termux/termux-packages#6846]
### Updates on this issue

* [2021.05.22] We have requested a VPS from FossHost. Request has been approved but it may take up to month for getting the server.
* [2021.06.01] Thanks to @Hero9909 for the providing container on his server. Repository setup is in progress. This issue will be closed once everything is ready.
* [2021.06.02] https://packages.termux.org is now up!
* [NOW] Waiting answer from FossHost regarding requested VPS. **No more suggestions accepted.**

***

Termux is looking for a new package hosting.

Previously we have used [Bintray](https://bintray.com) as our primary package hosting, providing support for Debian packages and REST API for repository management as well as donation of 60 TB/month of bandwidth for our project (thanks to JFrog). Though service is now [finished](https://jfrog.com/blog/into-the-sunset-bintray-jcenter-gocenter-and-chartcenter/) and we have to find another one.

**Write here if you wish to help us with hosting Termux repositories or have good suggestion of service which could fit our needs. See below on acceptable suggestions.**

## What we are looking for

Termux needs a primary/seed package host. It is importantly to note that this is not a mirror. Maintainers expected to be granted full access for uploading packages on individual basis as well as possibility of configuring automatic package upload from CI. Depending on the host type, e.g. if VPS, we may need to install a custom software for building & signing the repository. Package repository will be exposed through https://termux.org/ and serve large amounts of data.

Here are suitable variants:
1. **VPS**, which is not shared with third-party projects.
  Best solution since such server expected to be administrated by @termux team, without any third-party sudoers.
2. **SaaS platform for package hosting**.
  Do not suggest https://cloudsmith.com/ or https://packagecloud.io/ - they were tried and considered as *unsuitable*: significant issue is support only of predefined Debian/Ubuntu distributions and thus impossibility of defining a custom `apt` distribution/component and another issue is inflexibility of Deb package processing (will reject `x86_64` packages as it "expects" them to be `amd64` instead). Bintray didn't have these problems.
3. **S3-like object storage**.
  Worst, but acceptable solution as in such case we will have to build repo on CI side which will imply some potential issues.

Regardless of host type, it should have these properties:
* Storage: 25 GB (at least)
* Bandwith: 60 TB/month or unlimited
* Content management over SSH/SFTP or REST API.

It also would be perfectly if service is free, but of course we know that suitable free services are rare thing ;) Services which are paid or require identity verification should be discussed with @fornwall. I'm not going to manage billing nor provide documents on my own.

**(!!!)** *Please do not suggest public shell servers and other shady services which have problems with sharing their (very limited) hardware resources and troublesome to trust. We need to be sure that service will be stable in long term and secure enough to give it our GPG key for signing repository.*

## What is being used currently

We have set of [mirrors](https://github.com/termux/termux-packages/wiki/Mirrors) currently used to deliver packages. The primary ones are https://grimler.se and IPFS network nodes.

~~Note that there is already a pending candidate for Termux primary hosting - https://dl.kcubeterm.me. It is managed by @kcubeterm, yet works currently only as mirror.~~

**Comment by amm834** [comment]
VPS or IPFS

**Comment by xeffyr** [comment]
> VPS or IPFS

Regarding VPS: I'm aware about existing providers, i.e. Hetzner, Digital Ocean, Vultr and others, so here question is not which provider to choose. However I'm not going to provide a one on my own (otherwise why my mirrors are on IPFS?).

About IPFS: it is not mentioned in section "What we are looking for" as potential variant and thus not suitable. Moreover, it is used now to provide mirrors. But it is not a permanent solution. Mirror is not a primary host and provide read-only access to clients.

The primary node, host where IPNS (IPFS Name System) keys are stored, is hosted on my PC meaning that repository is up as soon as I still available online. IPNS records expire in 12 hours and should be re-broadcasted at least once within this period.

**Comment by eljoest** [comment]
Potential sponsorship options:

- [Digital Ocean](https://www.digitalocean.com/open-source/)
-[fastly](https://www.fastly.com/open-source#contact-community) (CDN)
- [scaleway](https://www.scaleway.com/en/about-us/open-source-program/) (credits for their services)
- hetzner and leaseweb appear to have sponsorships as well, but I haven't found dedicated contact information.

**Comment by buttaface** [comment]
Why do you want to serve everything from one package host, ie all 60 TBs/month from one server? Fredrik can easily rent an adequate VPS for $10-20/month, which let's you run any software and handle 2-4 TBs/month, and various mirrors like FOSShost could serve the remaining 56-58 TBs/month. The only problem with mirrors is keeping them up to date, but we could have a script monitor them and make sure they're getting updated, and derank them in the `pkg` installer if they're slow to update.

**Comment by kcubeterm** [comment]
@buttaface I applied for VPS from fosshost and they approved our project. We will get a dedicated vps soon. They are aware of our requirements since I points this issue on fosshost IRC. RAM size is yet to be discussed but we have numbers of choice in terms of software like nexus, pulp3 even aptly etc. By the way. Could you please share your ssh pubkey.

**Comment by xeffyr** [comment]
> RAM size is yet to be discussed but we have numbers of choice in terms of software like nexus, pulp3 even aptly etc.

Here is resource stats on processing packages from main, unstable and science repositories - these 3 have the largest packages.

I guess 4 GB RAM will be ok, considering the peak usage was `965156 kbytes`.

```
Command being timed: "aptly publish repo termux-test termux-test"
User time (seconds): 849.26
System time (seconds): 109.75
Percent of CPU this job got: 109%
Elapsed (wall clock) time (h:mm:ss or m:ss): 14:33.64
Average shared text size (kbytes): 0
Average unshared data size (kbytes): 0
Average stack size (kbytes): 0
Average total size (kbytes): 0
Maximum resident set size (kbytes): 965156
Average resident set size (kbytes): 0
Major (requiring I/O) page faults: 4
Minor (reclaiming a frame) page faults: 3563988
Voluntary context switches: 16187830
Involuntary context switches: 5297442
Swaps: 0
File system inputs: 1729848
File system outputs: 2024832
Socket messages sent: 0
Socket messages received: 0
Signals delivered: 0
Page size (bytes): 4096
Exit status: 0
```

***

> Why do you want to serve everything from one package host, ie all 60 TBs/month from one server?

Only one host can be primary. For the rest, mirrors are already available.

**Comment by SommerEngineering** [comment]
Hetzner seems like a good partner to me. I can imagine that they might act as a sponsor. But also their commercial offer seems to fit pretty well: An AX41 NVME server costs about 40 euros (about $49) per month and offers 2x 512 GB NVMe SSD, 64 GB RAM, 100 GB backup storage, 1 GBit/s connection and unlimited traffic. This is then a barebone machine, real hardware, no virtualization, etc.

Link: https://www.hetzner.com/dedicated-rootserver/ax41-nvme?country=en

**Comment by librehat** [comment]
May I suggest OVH as a potential provider? They have many products including VPS, dedicated server etc. The plans of VPS for example don't limit the traffic (even for the cheapest plan, though the speed is throttled).
I also found they do have quite many sponsorships and partnerships with well-known open-source projects: https://www.ovh.co.uk/news/press/

**Comment by buttaface** [comment]
Note that on the "unmetered" connections, which I have experience pushing TBs of traffic through for months at a time, you will not see anything like peak capacity. If somebody pushed 1 GBit/s through that Hetzner server, you would theoretically get 1 Gbps * 3600 s/hour * 720 hours/month / 8 bits/byte * 1000 GB /TB = 324 TB/month. But that's not going to be a dedicated network uplink,   so your real throughput could be 1-10% of that, or 3 to 30 TBs/month, depending on how many other servers share that 1 Gbps uplink in the data center.

**Comment by Hero9909** [comment]
Im in a Team of 3, we rented a server some time ago. currently we are not using all our resources and it looks like we wont use them all in the future. So ive talked to one member and awaiting currently the response of the other. If all goes well, i would share a lxc container with 4 c , 16G of ram and 100G of nvme storage, nested support is available in case of docker. our traffic is unlimited, guaranteed 1G link and currently only around 3-4% in use. keep in mind that as @buttaface said we dont know the structure of the datacentre maybe we share our traffic via overallocation. ill give an update if everything is planned and ready or canceled.

**Comment by Hero9909** [comment]
Okay, ive got the okay. Next week we are done with our halfe-year-update. After that we could start.
Ill wait until someone of termux-team gets in touch with me as i need an accountable for that.

**Comment by xeffyr** [comment]
@Hero9909 That's all nice, however there are questions:
* How long-term this solution could be?
  We expect host to be available at least for 2 years, preferably more.
* Is it like a shared server and our repositories are expected to be packaged as Docker or LXC container?
  Shared server is definitely a cons, which makes hard to delegate repository signing procedure on it. Will have to do that on external CI.
* Would it be accessible over SSH?

For now a VPS from FossHost looks like optimal solution.

**Comment by xeffyr** [comment]
@kcubeterm Are there any answers from FossHost?

**Comment by Hero9909** [comment]
@xeffyr 
1. we are planning to keep this as long as we have an accountable for that
2. isnt direct clear to me, correctme if i got this wrong. 
we have full control over the server and offering a lxc container (specs see comment before) with the distro of your choice.
From there any administration is done via ssh just like a vps.
3. yep 

if you like we could talk via discord or ts3 to get a faster q&a talk

**Comment by kcubeterm** [comment]
Build is in queue. It may take upto a month. They are really working hard to clear their backlog. I hope they managed to get as far as possible.

**Comment by xeffyr** [comment]
> we have full control over the server and offering a lxc container (specs see comment before) with the distro of your choice.
From there any administration is done via ssh just like a vps.

Thanks, will consider this as potential solution.

> Build is in queue. It may take upto a month.

Ok, till that continuing looking for other suggestions, in case FossHost will unable to provide it for some reason.

**Comment by librehat** [comment]
Just saw Oracle Cloud "Always Free" tier, would that be an option: https://www.oracle.com/uk/cloud/free/

**Comment by xeffyr** [comment]
@librehat https://free-for.dev/ - please don't re-provide solutions shown on this page.

Free tier is not same as free service. As stated in https://github.com/termux/termux-packages/issues/6846#issue-897890960:
> Services which are paid or require identity verification should be discussed with @fornwall.

Also Oracle Cloud has a quite problematic registration, I never was able to complete it (have tried 20+ times during several months) for various issues (one of: no email sent, inactive buttons when credit card verification passed, sometimes credit card verification fails, etc).


<details>
<summary>An example of issue during registration</summary>
Registration page shows that credit card verification has been successfully completed. However I can't proceed because confirmation checkbox is locked and button for beginning free tier is inactive.

<img src="https://user-images.githubusercontent.com/25881154/119821601-6b3a8c80-befb-11eb-8677-7a95ae67321c.png"/>
</details>

**Comment by librehat** [comment]
I am able to register and provision some running VM instances FYI.

But you're right, it requires credit card verification so it needs @fornwall

**Comment by xeffyr** [comment]
> I am able to register and provision some running VM instances FYI.

So what about egress speed? Many Oracle Cloud free tier reviews consider it as "low" and https://free-for.dev/ specifies it as 5mbps. Is that true?

**Comment by librehat** [comment]
> So what about egress speed? Many Oracle Cloud free tier reviews consider it as "low" and https://free-for.dev/ specifies it as 5mbps. Is that true?
Not 5Mbps. The AMD instances are limited to 0.48Gbps. The Ampere ARM instances are 1 ~ 4 Gbps depending on the configuration. Ampere instances are new to the Always Free Tier (since one or two days ago only) so maybe that's not updated on free-for.dev yet

**Comment by vhqtvn** [comment]
I use github release to host packages, and coded a simple repo which redirect deb download to github assets.

https://github.com/vhqtvn/termux-packages/releases/tag/repos-02498ab7d3ae890ea667f06e50cbda0e46c98751

**Comment by xeffyr** [comment]
On https://github.com/termux/termux-packages/issues/6846#issue-897890960 I have provided acceptable solutions. This even not the third one which is
> 3. Some kind of file hosting, preferably something similar to Amazon S3.

If you know what is Amazon S3, you will understand what kind of file hosting is expected. But VPS is still preferred.

So:
* How redirects are configured? Do you have any external server for metadata? **- We don't have any server now, which is why this issue is opened.**
* Could the packages be easily submitted from CI? As I understand, your solution is not suitable for `aptly` or other repository management software.
* If using CI, how concurrent uploads could be handled?

`termux-packages` is managed by multiple people which should be able to submit any amount of packages at any time and without interaction from my side.

*Hosting solution must be viable for setting up of F-Droid repository which would be done later as part of finishing https://github.com/termux/termux-app/tree/android-10.*

**Comment by vhqtvn** [comment]
> On [#6846 (comment)](https://github.com/termux/termux-packages/issues/6846#issue-897890960) I have provided acceptable solutions. This even not the third one which is
> 
> > 1. Some kind of file hosting, preferably something similar to Amazon S3.
> 
> If you know what is Amazon S3, you will understand what kind of file hosting is expected. But VPS is still preferred.

Yes I think this kind of storage is quite similar to S3, a place to put file, and obtain a link (not dirrect but still acceptable).

> So:
> 
> * How redirects are configured? Do you have any external server for metadata? **- We don't have any server now, which is why this issue is opened.**

Yes i'm using my server for metadata. But thinking about it, as files on release can be updated, we can even host the full repo (metadata + debs) on github release.

> * Could the packages be easily submitted from CI? As I understand, your solution is not suitable for `aptly` or other repository management software.
> * If using CI, how concurrent uploads could be handled?

Yes i'm using CI, modified from your current workflow.
- CI uploaded zip of tar of debs to server.
- Server loop through the uploaded debs:
  - calculate metadata and save using a key-value storage (key is hash of package and version) - currently it's just a folder with many hashes inside.
  - upload the deb to corresponding release (i'm using "repo-{sha}" as tag name) and queue repo reindex
- Reindex metadata using above key-value storage and rebuild the repo.

If we use github as repo we just need to flatten the repo struct and upload the metadata to github.

> `termux-packages` is managed by multiple people which should be able to submit any amount of packages at any time and without interaction from my side.
> 
> _Hosting solution must be viable for setting up of F-Droid repository which would be done later as part of finishing https://github.com/termux/termux-app/tree/android-10._

I'm not experience with F-Droid, but all the flow i described above can be coded into a github action, key-value storage can be a new github repository, the CI just need to pull that repo, add the value, commit and push. There wont be conflicts due to hash properties.

Regenerating metadata can upload result into a specific release, and that release is the repo root. We may need to code this into a CI's cron job instead to prevent overlapping run.

**Comment by xeffyr** [comment]
> Yes I think this kind of storage is quite similar to S3, a place to put file

No, it isn't. Storage must provide S3-compatible API, so `aptly` can work with it.

> we can even host the full repo (metadata + debs) on github release.

Releases do not support directories. Patching the `apt` is not an option.

> I'm not experience with F-Droid, but all the flow i described above can be coded into a github action, key-value storage can be a new github repository, the CI just need to pull that repo, add the value, commit and push. There wont be conflicts due to hash properties.

Too much effort for simple task.

**Comment by xeffyr** [comment]
*If you can't suggest a suitable hosting service or offer a ready-for-use VPS, preferably free of charge, or suggested hosting solution is worse than current one we have (existing Termux mirrors hosted by volunteers + IPFS), please avoid posting it there.*

https://github.com/termux/termux-packages/issues/6846#issue-897890960 describes which type of hosting we are looking for.

Please, do not suggest:
* Public shell servers *- some are good, but admins do not like to give a significant resource share and also such services are unreliable and totally insecure.*
* Abusing service demo/free plans *- we do not want termination of our accounts, right?*
* Using services which are not designed for our particular use case *- Potentially leads to increased maintenance effort, may lead to account termination if admins suspect ToS violation.*

I thought the last two points are obvious, considering that we are looking for stable long term hosting.

**Comment by RazyRabbit** [comment]
Why not to use Torrent to serve the packages?

You guys could serve psckages peer-to-peer with the App

@xeffyr

**Comment by xeffyr** [comment]
@RazyRabbit

As soon as we don't have a primary server, this won't be possible. Also torrents unsuitable for many technical reasons. As they are designed to deliver static content only, it would be problematic to deliver updates. Torrents solve only one issue - bandwidth. We already have mirrors for that.

In https://github.com/termux/termux-packages/issues/6846#issue-897890960 I have explained what type of hosting we are looking for, with storage and bandwidth requirements and preferred access method. Please don't suggest anything that doesn't meet provided criteria.

Example of good suggestion: https://fosshost.org/
Example of bad suggestion: https://pages.github.com/

**Comment by MC45892** [comment]
You could use GitHub pages

GitHub pages is hosted 24/7 and it free! if you want a custom domain you can use it for free you can buy a domain and connect to it for free or use gitlab pages or heroku

**Comment by MC45892** [comment]
> You could use GitHub pages
> 
> GitHub pages is hosted 24/7 and it free! if you want a custom domain you can use it for free you can buy a domain and connect to it for free or use gitlab pages or heroku

Also heroku only allows 5 apps for unverified users if you want to get unlimited apps then I think you have to pay for it

**Comment by selurvedu** [comment]
@MC45892 you literally suggested github pages right below xeffyr's comment stating that GitHub Pages are a bad suggestion. 🤦

**Comment by iBug** [comment]
Is it possible to further reduce bandwidth estimation (from 60 TB/mo) by *actively* prompting users to choose a mirror, or even do that automatically? Like CentOS does this with `fastestmirror` yum plugin.

**Comment by x0b** [comment]
> Is it possible to further reduce bandwidth estimation (from 60 TB/mo) by actively prompting users to choose a mirror, or even do that automatically?

With something like https://github.com/etix/mirrorbits mirror should work without user configuration, which is always a win.

Honestly, this issue is looking for unicorn - pushing that much traffic from a single machine/service with other requirements hinting at a beefy dedicated root server is going to cost serious money. Leaning more heavily onto mirrors as *primary* distribution method (instead of just as _backup_) is imho the only way to move forward (unless we get sponsored resources, like we may receive from fosshost).

**Comment by Hero9909** [comment]
i wont support this as i use mobile data, what others might do althought.  therfore i think a opt-in solution is more user friendly. 
anyway a seed server is needed in any way, even if this user mirrors are used as there are old clients that need the old structure. Futher i  think this needs serious work to set up, considering a mixed solution to migrate most clients. 

anyway im ready to get the container running @xeffyr. if you like i can setup a subdomain for this container. 

what i need to launch:
distro choice
subdomain name choice
public key

no hurry i just wanted to leave this here.

**Comment by xeffyr** [comment]
@Hero9909 I'm ready to start with your hosting suggestion. Here are the details:

* Distribution: Ubuntu 20.04
* Subdomain: `packages.termux.org`
* Public key:
  ```
  ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDQKamTQqcujLZo4w0lQHh4GYE+ecMHBMYPk2SkQHq8gmwApO9vs07H09idSBVRqYBK6mMg3T9GJgqGFCcLzK4GdV2QcJGIuqxyHKGxWE2uymSvY0a3JEfYhc4dLKHopdvk0n+1RRa3JlvoBvPz5U3nCsky9Du5BtvcqFLphc8/pGaX78WmqLUVpql8Dav5RyBQkDGeAlcfNqsrOBNkLl/qHyv0g36F+O8M/B32eSlQs58hs8ecG90SXJkO1KjxcH/KiEJckfD7YM6FCHfgvzXKpd1naoo75TRHgK1DSjEWDmpedpSIygJKvocZ91rQemJ3P6G/Zy+78emVdmrBuFzEWRouhky0yN8hRzhfcdsfzPCOYeHI3JsmXQ7txb+iDiAP2IaA1EMX2ngr704O6syGbeoAtBnOZY9bM6qL05ruVsEcZvPL1YL1IcNwdq7f/MVbLaUL4BhHIJ91attff8EFPu5mB1Y+QjLRnobB7GgpGo7NM9oivJvgki5DNmNJi3vJEMmv1UKDRZbpnZDmdtNywg0zTIlz09DI6u+IJeXPrhQSzlmPm2aT3QwQXv9CMzkQfuK3kmLJiUWZAvZVjH7amlIfJN4r08bHuBUXckpKKjMVzh/EOCzNfsR49owyBej6CmN/aWqmuE3NyPJIYatS9/jU2ItvOzjAa94g/KUlUw==
  ```

Probably @Grimler91 and @kcubeterm would like to provide their SSH pubkeys as well.

Note that host choice may be re-considered if FossHost will provide us a VPS. But for now I guess a container on shared server would be fine, since it will require some time to establish automatic package submissions from scratch and ensure whether it works without issues.

**Comment by Hero9909** [comment]
Okay @xeffyr container is running, please contact me after reading the readme file in your home directory.
Host: tanisorhost.cosmos-ink.net
Port: 2207
User: xeffyr

**Comment by kcubeterm** [comment]
@Hero9909 
Here's my ssh pubkey 
```
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQDXmC2QjmDYeNa26I05Tgj8ElunNMibYKCP5A0iky77ZInEmi6WYU96Vq158+dZr0Tbr7wj1YuZ030oHrcp2i2hUeimc0/YGRTCh7PRuIDho2K+miw/5/KRpdIqDL+YT1A7fg+7qnbt6CfdGrsXJVwvafzQ3Bte3Rc3aenGeYJrm0T8LfK6kwLxtlZeLzbUuSIaf//jZthEWvJvEUYjYX/NH1tHDG5z4rYLXfrsMLwyTDfGOwFEPM1KJR5zl9BXeP9KlxmobTciG10ZNrOWPV5vfxySMcrL4QfsUm+daJHO5cDZRlXx7auUJ6rUId7TbJiTq24vnoNEx5XRVYNbYZVrbT90+xhaV7fR33rF08pNkYGgwRwH8sVo4E5MCsNiJVIoCHbu8oZVUuLFkSKOOLajGUIkG+SvRY6malWieT1uCNjDRRMNsA8l5BdcX6be+0j5hPbej2L0huh+xujvNLlA87TyKxROhinZzy0fBmIR9C+D9jRUopL+gdmdQVHcUQU=
```

-------
I was working on GitHub release as well. almost finished. Now I keep it aside as we have now a server for hosting primary repository. Who knows it may need in future ;)

**Comment by Hero9909** [comment]
Okay @kcubeterm ssh key added, please contact me after reading the readme file in your home directory.
Host: tanisorhost.cosmos-ink.net
Port: 2207
User: kcubeterm

**Comment by librehat** [comment]
Now we're going to have a primary host but we may still want to consider choosing the fastest mirror automatically when running `pkg` so the load is more distributed?

(I'm configuring an ARM instance on Oracle Cloud in the UK data centre. It has > 40G free space, 1 Gbps bandwidth and 10 TB limit every month)

**Comment by xeffyr** [comment]
@Hero9909 I have problems accessing the host.
```
ssh -p 2207 xeffyr@tanisorhost.cosmos-ink.net
ssh: connect to host tanisorhost.cosmos-ink.net port 2207: Network is unreachable
```

Ping shows this output:
```
ping tanisorhost.cosmos-ink.net
PING tanisorhost.cosmos-ink.net (95.217.104.7) 56(84) bytes of data.
64 bytes from static.7.104.217.95.clients.your-server.de (95.217.104.7): icmp_seq=1 ttl=57 time=42.2 ms
```

**Comment by kcubeterm** [comment]
I also couldn't
```
ssh kcubeterm@tanisorhost.cosmos-ink.net -p 2207
ssh: connect to host tanisorhost.cosmos-ink.net port 2207: Connection refused
```

**Comment by Hero9909** [comment]
hm @xeffyr  use 95.217.104.7 as fallback, can confirm that the requests not going throught. im looking into this

**Comment by Hero9909** [comment]
@xeffyr @kcubeterm fixed

**Comment by xeffyr** [comment]
@Hero9909 I can confirm that now can access it.

Is web server (nginx) is entirely managed by you or you can proxy it inside the container so we will have own config?

**Comment by Grimler91** [comment]
@Hero9909 could you add a user for 
```
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQDXFvKMueZ481gw2nXj7QBME8O6LexKh+rFjzP8B7iVeuxbLgbQiWs/6WBMUfdV0a784KOaEJcNz7imarNYbtVkthTAl+2I3DVYMEx7WqRdgEzoUsgK7OmCK9odyeQfkQ2ZT6dnPM8lxGSHw/jhBan1oDPxnjAi70zmFuxI3mz4MSlkuAeyBgV1mHRLoVbaIcu1dKbhDk+eWod2kLMLRy/Pkc8belcMS+o0Rn3X1ZpcaZ9dCNWl8e6zGmlLLcia+xa8ePUOUzy43oXYUjhKkmOXU5onmSqgoAS6l6rQHn6J3lWeMb29ifD0SwbdZ3aCfrPeZbebv7WD2olzBJ8WMT+/6j053itqMTm7SH33RlKOW01m7La+hM1ZZ2+fWoir+BIXpzpLbqGP/TddsN9/iUM6lr6KSI/lCzS28JjBkBeq7onG+I9cSlPQ4dAPnY9Eg/pPgZ+X76dHqhXwvIKWaT5rOkA8clvir6bR4bYCrjGKmDsb9kou5j4P/zQDd8hdUw8= grimler
```
as well? Thanks!

**Comment by Hero9909** [comment]
ive already talked to another member about this, we are be abled to setup a seperate server/proxy for the termux.org domain, but well need the certificates, another way is to use the direct mapped ipv6, but this wont work for all users as ipv6 isnt as common as it should be.

**Comment by Hero9909** [comment]
@Grimler91 Done, please contact me after reading the readme file in your home directory.
Host: tanisorhost.cosmos-ink.net
Port: 2207
User: grimler91

**Comment by xeffyr** [comment]
We don't have certificates, but can use Let's Encrypt for obtaining them. This will require automated renew since they expire in 90 days. Other variant will be use of CloudFlare which will also act as proxy (and CDN) and let IPv4 users access the server.

It is expected that container will run these 2 services which should be accessible outside:
* Apt repository (assume just file server)
* Aptly [API server](https://www.aptly.info/doc/api/) which will require reverse proxy and authentication.

So that's why I'm asking whether we will be able to manage the config on our own.

**Comment by Hero9909** [comment]
im working on a solution, for security reasons i would like to migrate futher detail discussions to a non public place, is telegram okay for you?

**Comment by xeffyr** [comment]
> is telegram okay for you?

Okay

**Comment by Grimler91** [comment]
@Hero9909 I am getting permission denied, could you check if the pub key was added correctly?

(Fingerprint of key I posted here matches the private key, so should be correct)

**Comment by Hero9909** [comment]
try again.

**Comment by xeffyr** [comment]
https://packages.termux.org/ is now set up and works fine, so I'm closing this issue.

***

@kcubeterm, @Grimler91, @a1batross Suggesting to reconfigure your mirrors for cloning https://packages.termux.org/.

Main repository:
```
deb https://packages.termux.org/apt/termux-main/ stable main
```

Games repository:
```
deb https://packages.termux.org/apt/termux-games/ games stable
```

Root repository:
```
deb https://packages.termux.org/apt/termux-root/ root stable
```

Science repository:
```
deb https://packages.termux.org/apt/termux-science/ science stable
```

Unstable repository:
```
deb https://packages.termux.org/apt/termux-unstable/ unstable main
```

X11 repository:
```
deb https://packages.termux.org/apt/termux-x11/ x11 main
```

As you can see, paths there different. Though you can try to map them to
```
termux-packages-24
game-packages-24
termux-root-packages-24
science-packages-24
unstable-packages
x11-packages
```
to maintain compatibility. If you decide a fresh configuration with new paths, notify me and I will update `termux-tools` package and information at https://github.com/termux/termux-packages/wiki/Mirrors.

**Comment by agnostic-apollo** [comment]
Thanks a lot @xeffyr for setting this up and @Hero9909 for providing the container.

**Comment by a1batross** [comment]
@xeffyr Thank you! Changed paths and added symlinks to maintain compatibility on my side.

**Comment by truboxl** [comment]
@xeffyr can you update https://github.com/termux/termux-docker as well? Thanks!

**Comment by iBug** [comment]
Do you provide Rsync service for downstream mirror sites?

**Comment by xeffyr** [comment]
No, we don't provide rsync currently.

**Comment by iBug** [comment]
How long will the IPFS links stay around? Might need some time before I can update our mirror config.

**Comment by kcubeterm** [comment]
Until any single nodes have that content. If all nodes goes down and neither gateway has cached version. Links won't work. Right now only @xeffyr node is serving, I stopped all of my nodes.

**Comment by selurvedu** [comment]
I am very glad that my suggestion of Fosshost worked. Thank you all for your efforts!

**Comment by buttaface** [comment]
@xeffyr, now that we have a dedicated package server again, can we turn [the automated update checker back on on the github CI](https://github.com/termux/termux-packages/blob/master/scripts/updates/check-updates.sh)? I thought that really helped us keep github projects up to date.

**Comment by landfillbaby** [comment]
maybe we should make a list of packages that aren't automatically updated too so we have an explicit list of what to check manually

**Comment by buttaface** [comment]
[There is already a list of mostly the inverse](https://github.com/termux/termux-packages/blob/master/scripts/updates/github-projects.txt), but it's not complete.

**Comment by iBug** [comment]
Hi @xeffyr, will you update the repository wiki?

I've updated [mirrors.ustc.edu.cn](https://mirrors.ustc.edu.cn/termux/apt/) to sync from packages.termux.org, with the new directory layout and backwards compatibility with the old ones from Bintray era (just some more symbolic links).

Our current URLs are as follows. Use the edit button to retrieve Markdown source from this comment.

---



## Mirrors by [University of Science and Technology of China, Linux User Group](https://lug.ustc.edu.cn/)

Mirror for Chinese users for better ping and download speed.

|Repository|sources.list entry                                               |
|:---------|:----------------------------------------------------------------|
|[Main](https://github.com/termux/termux-packages)      |`deb https://mirrors.ustc.edu.cn/termux/apt/termux-main/ stable main`|
|[Games](https://github.com/termux/game-packages)     |`deb https://mirrors.ustc.edu.cn/termux/apt/termux-games/ games stable` |
|[Root](https://github.com/termux/termux-root-packages)      |`deb https://mirrors.ustc.edu.cn/termux/apt/termux-root/ root stable`|
|[Science](https://github.com/termux/science-packages)   |`deb https://mirrors.ustc.edu.cn/termux/apt/termux-science/ science stable`|
|[Unstable](https://github.com/termux/unstable-packages)  |`deb https://mirrors.ustc.edu.cn/termux/apt/termux-unstable/ unstable main`|
|[X11](https://github.com/termux/x11-packages)       |`deb https://mirrors.ustc.edu.cn/termux/apt/termux-x11/ x11 main`|

**Comment by xeffyr** [comment]
@iBug Thanks, repository wiki is updated now.

**Comment by iBug** [comment]
@xeffyr Can you update the wiki again for accuracy? There are two points missing here:

- I changed the title (Mirror by ...) for better grammar
- We focus on mainland China so I prefer that this line be there:

    > Mirror for Chinese users for better ping and download speed.

You can also update `termux-tools` to include our mirror. A short, suitable name would be `Mirrors by USTC`.

**Comment by NatoBoram** [comment]
I was rehosting the termux packages on IPFS (https://github.com/ipfs/go-ipfs/issues/4435#issuecomment-793154227), but it looks like the IPNS stopped resolving. I'm glad that Termux found a suitable hosting service!
